src/ogdf/basic/GraphCopy.cpp

```diff
@@ -112,202 +91,150 @@ void GraphCopySimple::delEdge(edge e) {
+	for (edge original_edge : original().edges) {
+		for (edge other_edge : other.m_eCopy[original_edge]) {
+			edge my_edge = edgeMap[other_edge];
+			m_eOrig[my_edge] = original_edge;
+			m_eIterator[my_edge] = m_eCopy[original_edge].pushBack(my_edge);
 		}
 	}
-}
-
-void GraphCopy::init(const Graph& G) {
-	m_pGraph = &G;
-
-	EdgeArray<edge> eCopy;
-	Graph::assign(G, m_vCopy, eCopy);
-
-	m_vOrig.init(*this, nullptr);
-	m_eOrig.init(*this, nullptr);
-	m_eCopy.init(G);
-	m_eIterator.init(*this, nullptr);
-
-	for (node v : G.nodes) {
-		m_vOrig[m_vCopy[v]] = v;
-	}
-
-	for (edge e : G.edges) {
-		m_eIterator[eCopy[e]] = m_eCopy[e].pushBack(eCopy[e]);
-		m_eOrig[eCopy[e]] = e;
-	}
-}
-
-void GraphCopy::createEmpty(const Graph& G) {
-	m_pGraph = &G;
+	return *this;
+};
 
+void GraphCopy::setOriginalGraph(const Graph* G) {
+	m_pGraph = G;
+	m_vOrig.init(this, nullptr);
+	m_eOrig.init(this, nullptr);
 	m_vCopy.init(G, nullptr);
 	m_eCopy.init(G);
-	m_vOrig.init(*this, nullptr);
-	m_eOrig.init(*this, nullptr);
-	m_eIterator.init(*this, nullptr);
-}
-
-void GraphCopy::initByCC(const CCsInfo& info, int cc, EdgeArray<edge>& eCopy) {
-	eCopy.init(*m_pGraph);
-
-	m_vCopy.init(*m_pGraph, nullptr);
-	m_eCopy.init(*m_pGraph);
-	Graph::constructInitByCC(info, cc, m_vCopy, eCopy);
-
-	for (int i = info.startNode(cc); i < info.stopNode(cc); ++i) {
-		node v = info.v(i);
-		m_vOrig[m_vCopy[v]] = v;
-	}
-
-	for (int i = info.startEdge(cc); i < info.stopEdge(cc); ++i) {
-		edge e = info.e(i);
-		m_eIterator[eCopy[e]] = m_eCopy[e].pushBack(eCopy[e]);
-		m_eOrig[eCopy[e]] = e;
-	}
-
-#ifdef OGDF_HEAVY_DEBUG
-	consistencyCheck();
-#endif
+	m_eIterator.init(this, nullptr);
 }
 
-void GraphCopy::initByNodes(const List<node>& origNodes, EdgeArray<edge>& eCopy) {
-#ifdef OGDF_DEBUG
-	ArrayBuffer<node> stack;
-	List<node> copyList(origNodes);
-	while (!copyList.empty()) {
-		stack.push(copyList.popFrontRet());
-		node v;
-		while (!stack.empty()) {
-			v = stack.popRet();
-			for (adjEntry adj : v->adjEntries) {
-				node w = adj->twinNode();
-				OGDF_ASSERT(origNodes.search(w).valid());
-				auto it = copyList.search(w);
-				if (it.valid()) {
-					stack.push(w);
-					copyList.del(it);
-				}
-			}
+void GraphCopySimple::setOriginalEmbedding() {
+	List<adjEntry> newAdjOrder;
+	for (node v : getOriginalGraph()->nodes) {
+		node u = copy(v);
+		if (u == nullptr) {
+			continue;
 		}
-	}
-#endif
-
-	m_vCopy.init(*m_pGraph, nullptr);
-	m_eCopy.init(*m_pGraph);
-	Graph::constructInitByNodes(*m_pGraph, origNodes, m_vCopy, eCopy);
-
-	for (node v : origNodes) {
-		m_vOrig[m_vCopy[v]] = v;
-
+		newAdjOrder.clear();
+		// newAdjOrder.reserve(v->degree());
+		// add from original according to their order
 		for (adjEntry adj : v->adjEntries) {
-			if ((adj->index() & 1) == 0) {
-				edge e = adj->theEdge();
-				m_eIterator[eCopy[e]] = m_eCopy[e].pushBack(eCopy[e]);
-				m_eOrig[eCopy[e]] = e;
+			edge e = copy(adj->theEdge());
+			if (e != nullptr && e->isIncident(u)) {
```

Comment:
I remember initially using the same `setOriginalEmbedding` code for all versions of `GraphCopyBase`, but this broke in `GraphCopy`. The issue was probably that `copy(adj)` sometimes returns adjEntries with `copy(adj)->theNode() != copy(adj->theNode())` (maybe because it per its docs doesn't care about reversed edges?). My simple solution to this was splitting the method into [two versions](
https://github.com/ogdf/ogdf/commit/60f9a8ecc893c3443e9136327759ca0301315534#diff-90c54bd7a34123d3157346d20b16a48e9582030899d09a70beb49d4bd1014343) and simply treating such edges as new/unembedded ones whenever this happens for `GraphCopy`. We thereby still support strictly more use-cases than before, but this could probably be further improved. Maybe checking for reversed edges is already enough, but I'd leave that to someone with more experience with the workings of the non-simple `GraphCopy`.